### DISCUSSION TITLE
expanding keyboards problem

### DISCUSSION THREAD

**Discussion by fabiomancini67** [GrandOrgue/GoOdf#103] [General]
I'm trying to modify an odf that I created some times ago, it's a 2 manuals organ with 56 notes keyboards and I want to expand them to 61. The ODF used ranks. I'm using GoOdf 0.10.0 for windows. 

I follow these steps:
- Manuals section -  Keyboards section
I modify number of logical keys from 56 to 61
I modify number of accessible keys from 56 to 61
Then for each Stop I modify number of accessible pipes from 56 to 61
- Ranks section
I modify number of logical pipes from 56 to 61

At this point I go to pipe 56, right click - copypipeoffset to.... I apply it from 57 to 61 and automatically also the right pitch tuning is applied (very cool feature!).

In grandorgue the new keys doesn't sound except the key 57. I also tried to borrowing the pipes instead of copypipeoffset to.... with the same result.

If I open the odf with a text editor and I go in the STOP section I can see that a line is added for each modified stop: RankXXXPipeCount=56 
If I manually modify this line and I change 56 to 61 everything is ok

Also I noted that in the Edit pipe notes, the last added pipe has the "percussive" flagged on YES, but it seems that it's not really applied.

**Comment by larspalo** [comment]
> If I open the odf with a text editor and I go in the STOP section I can see that a line is added for each modified stop: RankXXXPipeCount=56
If I manually modify this line and I change 56 to 61 everything is ok

Have you watched https://youtu.be/hzpuzJnttWs?feature=shared&t=752? At around 12.30 and a bit forward there showcases this situation. The quick fix is just removing the referenced rank from the stop and adding it back to update the pipe count.

There's a subtle reason why I've not implemented updating the pipe count in the stop that references a rank immediately when changing the pipe numbers in the rank. That's simply because it's perfectly valid, and a real important use case for ranks as such, that a stop might only reference parts of a rank. There could for instance be a single rank that had some 85 pipes that would be used in three different stops at 16', 8' and 4' pitch.

Hope that explains it somewhat.

> Also I noted that in the Edit pipe notes, the last added pipe has the "percussive" flagged on YES, but it seems that it's not really applied.

I've currently got no explanation for that, it's nothing I've ever encountered myself so far. Do you have steps to reproduce it reliably?

**Comment by Bumblebee001** [comment]
I'm having issues in this regards too.... to the extent that I manually created all the missing keys missing from 57 to 61 by pitch shifting the last 5 recorded notes, using an audio editor, re-doing markers and loops as if the samples were real recordings themselves. The issues I am facing is how to get the suboctave and octave couplers to sound pipes that are outside of the 61 key spectrum of the organ keyboard. Yes I did watch the screencast but got lost.... possibly my attention span is waning or I have early dementia, or something similar... I worked through the software and tried different settings. What's happening is that the suboctave coupler worked with C60 (middle C) and higher but fails in the first two octaves. Octave coupler seems to fail beyond the C96 (last note on the keyboard). I have set things like first midi note number=24, number of logical pipes=85, number of keys 85 (!!), first accessible pipe logical key number=1, first accessible pipe number=1 etc etc. I have edited the pipes at both extremes according -1200 and 1200 respectively and to include the following 11 pipes.... Flexible file loading option - tried that too....

Also I note that the expanded trees of the pipes show the references only for the attack samples but no releases at all. Why???

![Screenshot 2025-03-23 092318](https://github.com/user-attachments/assets/74b4e877-6f11-48cc-a5d5-da063725de3e)

(Other numbers/settings I used: first midi note=36, first accessible key=13 etc were not helpful either).

I'm getting confused also with what to select in the coupler dialogue - upward/downward, intermanual/ intramanual - couplers settings but I don't think this is important at this stage and I will have to try watching the screencast for these. 

And I thought I should be a veteran in creating odfs for GO by now!!!! 

Rather depressing as I'm starting to blame my age and general health for the failings!!!! I can still cook different good meals though, thank God!!!! :-D

**Reply by larspalo** [reply]
> Also I note that the expanded trees of the pipes show the references only for the attack samples but no releases at all. Why???

A reference is not a "copy" of another pipe - it's the _same_ pipe already existing and defined elsewhere (REF:MANUAL:STOP:PIPE numbers) so of course no additional attacks or releases would be shown. You'd have to look at the actual pipe that's referenced to see the attacks/releases.

Also, note that REF "pipes" cannot be re-pitched in the usual way, but you can reference already re-pitched pipes or reference them at other "pitches" but you cannot adjust fine re-tuning in the referencing "pipe" - that can only be done in the referenced pipe itself.

Basically, only use REF if the pipes are "borrowed" (organ term).

**Comment by Bumblebee001** [comment]
I went through the screencast again, took down notes and made my own schematic/ summary of the changes needed. I followed each step to the letter (or should I say "number"?). What was relevant to me was Stop extension either side of the manuals

The difference with my organ is that I don't have Unison off stops.

Upon loading, I get several errors relating to temperament retuning by more than 600 cents.... for all notes of all the stops.

Secondly, I don't notice any change to the way the octave/ suboctave couplers should work. So basically the whole exercise did not work for me. What could be wrong?

**Reply by Bumblebee001** [reply]
Found the issue.... I set the manual details correctly for the wrong manual! 

Another senior moment! Grrrr!

The 600 cent errors for the edited stops have all cleared for the edited stops. What remains are related to stops I still have to apply the changes to.

**Comment by Bumblebee001** [comment]
I needed to do the same for a Gamba stop that starts at 048-C. It seems that I succeeded in applying the same methodology applying different values of course. What I ended up with is a stop that normally would not play the first octave. Even though I specified the accessible keys as starting from key 13, the lower octave keys plays nevertheless. Whilst they should sound when suboctave coupler is used, they should not sound at all when the keys are played. 

Any advice?  I tried first accessible pipe = 13 for the stop. I did not notice any change!!!

**Reply by larspalo** [reply]
Remember that each stop has a property like FirstAccessiblePipeLogicalKeyNumber that can be used to indicate that the Pipe001 of that stop is not at the normal lowest key of the manual.

**Reply by Bumblebee001** [reply]
I set that at 13 as I mentioned..... but the lowest octave now plays after adding the extension of that stop (with or without the suboctave on).

If I try to change that number (to 12, 14 or 25) it will **not** save!!!! Well.... it seems to in NotePad++ but the change is not seen in GoODF. See screenshot:

![Screenshot 2025-03-25 034213](https://github.com/user-attachments/assets/b0bbe85f-32b7-4a6d-910b-5c80fec4e89a)

**Comment by Bumblebee001** [comment]
Another problem: if one extends say the last 5 notes (recorded notes amount to 56 keys/notes) to bring the manual to 61 keys (pitch tuning each by 500, when it comes to extending the last octave as I have been doing for the octave coupler, the last 5 borrowed / "created" notes will not extend to complete the upper end of the 85 logical pipes.... if you see what I mean. The last 5 pipes (remain "dummies".

Any advice for this short of creating the notes with an audio editor.

**Reply by larspalo** [reply]
> Any advice for this short of creating the notes with an audio editor.

GO and GoOdf should allow from +1800 to -1800 PitchTuning adjustments, but it's almost certain that you'd get much better audio quality from re-pitching the samples in an audio editor.
